File: Minimum Size Subarray Sum ( positive integer )/Javascript/solution_3.js
// https://leetcode.cn/problems/2VG8Kg/
// https://www.lintcode.com/problem/406

function minSubArrayLen(target, nums) {
  let nums_len = nums.length;

  // 异常处理
  if (!nums || !nums_len) {
    return 0;
  }

  // 获取前缀和
  let prefix_sum_arr = prefix_sum(nums);

  // 最小值
  let min_len = Infinity;

  for (let start = 0; start < nums_len; start++) {
    let end = find_end_pos(prefix_sum_arr, start, target);
    if (prefix_sum_arr[end + 1] - prefix_sum_arr[start] >= target) {
      let len = end - start + 1;
      min_len = Math.min(min_len, len);
    }
  }

  return min_len === Infinity ? 0 : min_len;
}

function find_end_pos(prefix_sum, start, target) {
  let left = start;
  let right = prefix_sum.length - 2;

  while (left + 1 < right) {
    let mid = left + ~~((right - left) / 2);

    if (prefix_sum[mid + 1] - prefix_sum[start] >= target) {
      right = mid;
    } else {
      left = mid;
    }
  }

  if (prefix_sum[left + 1] - prefix_sum[start] >= target) {
    return left;
  }

  return right;
}

function prefix_sum(nums) {
  let prefix_sum_arr = [];
  prefix_sum_arr[0] = 0;

  for (let i = 0, len = nums.length; i < len; i++) {
    prefix_sum_arr[i + 1] = prefix_sum_arr[i] + nums[i];
  }

  return prefix_sum_arr;
}

// test cases
let target = 7;
let nums = [2, 3, 1, 2, 4, 3];
console.log(minSubArrayLen(target, nums));

File: Minimum Size Subarray Sum ( positive integer )/Rust/src/solution_3.rs
// https://leetcode.cn/problems/2VG8Kg/
// https://www.lintcode.com/problem/406

pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
    let nums_len = nums.len();

    // 异常处理
    if nums_len == 0 {
        return 0;
    }

    let mut min_len = i32::MAX;

    let mut end = 0;
    let mut subarray_sum = 0;

    for start in 0..nums_len {
        while end < nums_len && subarray_sum < target {
            subarray_sum = subarray_sum + nums[end];
            end = end + 1;
        }

        if subarray_sum >= target {
            let len = (end - start) as i32;
            min_len = if min_len > len { len } else { min_len }
        }

        subarray_sum = subarray_sum - nums[start];
    }

    if min_len == i32::MAX { 0 } else { min_len }
}